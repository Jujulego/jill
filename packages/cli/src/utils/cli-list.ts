import chalk from 'chalk';

// Class
export class CliList {
  // Attributes
  private _headers?: string[];
  private readonly _data: string[][] = [];
  private readonly _columns: number[] = [];

  // Methods
  private _updateColumns(data: string[]): void {
    // Add missing columns
    while (data.length > this._columns.length) {
      this._columns.push(0);
    }

    // Update columns
    for (let i = 0; i < data.length; ++i) {
      this._columns[i] = Math.max(this._columns[i], data[i].length);
    }
  }

  private _formatValue(value: string, idx: number): string {
    return value + ' '.repeat(this._columns[idx] - value.length);
  }
  
  setHeaders(...headers: string[]): void {
    this._updateColumns(headers);
    this._headers = headers;
  }
  
  add(data: string[]): void {
    this._updateColumns(data);
    this._data.push(data);
  }

  *lines(): Generator<string, void> {
    if (this._headers) {
      yield this._headers.map((v, i) => chalk.bold(this._formatValue(v, i)))
          .join('  ');
    }

    for (const data of this._data) {
      yield data.map((v, i) => this._formatValue(v, i))
        .join('  ');
    }
  }

  // Properties
  get columns(): number[] {
    return this._columns;
  }

  get headers(): string[] {
    return this._headers || [];
  }

  get data(): string[][] {
    return this._data;
  }
}