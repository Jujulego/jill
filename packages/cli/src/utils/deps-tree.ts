import { Workspace } from '@jujulego/jill-core';
import chalk from 'chalk';

// Functions
async function printTree(wks: Workspace, level: string, dev: boolean, printed: Set<string>) {
  const workspaces: [Workspace, boolean][] = [];

  for await (const dep of wks.dependencies()) {
    workspaces.push([dep, dev]);
  }

  for await (const dep of wks.devDependencies()) {
    workspaces.push([dep, true]);
  }

  for (let i = 0; i < workspaces.length; ++i) {
    const [dep, isDev] = workspaces[i];
    const isPrinted = printed.has(dep.cwd);
    const isLast = i === workspaces.length - 1;

    // Format
    let name = dep.name;
    if (isDev) name = chalk.grey(`${name} (dev)`);
    if (isPrinted) name = chalk.italic(name);

    const branchFmt = dev ? chalk.grey : (s: string) => s;

    console.log(`${level}${branchFmt(`${isLast ? '└' : '├' }─ `)}${name}`);

    // Print deps of dep
    if (!isPrinted) {
      printed.add(dep.cwd);
      await printTree(dep, level + (isLast ? '   ' : branchFmt('│  ')), isDev, printed);
    }
  }
}

export async function printDepsTree(wks: Workspace): Promise<void> {
  console.log(chalk`{bold Dependencies:}`);
  const printed = new Set([wks.name]);

  await printTree(wks, '', false, printed);

  if (printed.size === 1) {
    console.log(chalk.grey(`   No dependencies for ${wks.name}`));
  }
}