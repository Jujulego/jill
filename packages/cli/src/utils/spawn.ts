import * as cp from 'child_process';

import { logger, OraLogger } from '../logger';

// Type
export interface SpawnResult {
  stdout: string[];
  stderr: string[];
}

export interface SpawnOptions {
  cwd?: string;
  env?: Record<string, string>;
  logger?: OraLogger;
}

// Utils
export function spawn(cmd: string, args: ReadonlyArray<string>, opts: SpawnOptions = {}): Promise<SpawnResult> {
  const log = opts.logger ?? logger;

  return new Promise<SpawnResult>((resolve, reject) => {
    log.debug(`Running ${[cmd, ...args].join(' ')}`);

    const proc = cp.spawn(cmd, args, {
      cwd: opts.cwd,
      shell: true,
      stdio: 'pipe',
      env: {
        ...process.env,
        ...opts.env
      }
    });

    // Gather result
    const res: SpawnResult = {
      stdout: [],
      stderr: []
    };

    proc.stdout.on('data', (msg: Buffer) => {
      res.stdout.push(...msg.toString('utf-8').replace(/\n$/, '').split('\n'));
    });

    proc.stderr.on('data', (msg: Buffer) => {
      res.stderr.push(...msg.toString('utf-8').replace(/\n$/, '').split('\n'));
    });

    // Emit result
    proc.on('close', (code) => {
      if (code) {
        reject(new Error(res.stderr[res.stderr.length - 1] || `${[cmd, ...args].join(' ')} failed`));
      } else {
        resolve(res);
      }
    });
  });
}