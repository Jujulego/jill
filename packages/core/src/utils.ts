import { streamEvents } from '@jujulego/event-tree';
import { SpawnTask, SpawnTaskStream } from '@jujulego/tasks';

// Utils
export async function* combine<T>(...generators: AsyncGenerator<T>[]): AsyncGenerator<T> {
  for (const gen of generators) {
    yield* gen;
  }
}

export async function *streamLines(task: SpawnTask, stream: SpawnTaskStream): AsyncGenerator<string> {
  // Abort
  const ctrl = new AbortController();

  task.subscribe('status.done', () => ctrl.abort());
  task.subscribe('status.failed', () => ctrl.abort());

  // Stream
  let current = '';

  try {
    for await (const chunk of streamEvents(task, `stream.${stream}`, { signal: ctrl.signal })) {
      const data = current + chunk.data.toString('utf-8');
      const lines = data.split(/\r?\n/g);

      current = lines.pop() ?? '';

      for (const line of lines) {
        yield line;
      }
    }
  } catch (err) {
    if (task.exitCode !== 0) {
      throw err;
    }

    if (current) {
      yield current;
    }
  }
}
