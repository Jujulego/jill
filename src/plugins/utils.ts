import { type Plugin } from './types';

// Plugin utils
export function definePlugin<T, U>(plugin: Plugin<T, U>): Plugin<T, U> {
  return plugin;
}

export function assertPlugin(obj: unknown, name: string): asserts obj is Plugin {
  if (!obj) {
    throw new Error(`Plugin ${name} is not a valid plugin. Default export is null or undefined`);
  }

  if (typeof obj !== 'object') {
    throw new Error(`Plugin ${name} is not a valid plugin. Default export is a ${typeof obj}`);
  }

  if (!('builder' in obj)) {
    throw new Error(`Plugin ${name} is not a valid plugin. Missing builder method in default export`);
  }
}
