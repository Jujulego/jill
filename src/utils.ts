import { streamEvents } from '@jujulego/event-tree';
import { SpawnTask, SpawnTaskStream } from '@jujulego/tasks';
import { Arguments, Argv, CommandModule } from 'yargs';
import { Awaitable } from './types';

// Stream utils
export async function* combine<T>(...generators: AsyncGenerator<T>[]): AsyncGenerator<T> {
  for (const gen of generators) {
    yield* gen;
  }
}

export async function *streamLines(task: SpawnTask, stream: SpawnTaskStream): AsyncGenerator<string> {
  // Abort
  const ctrl = new AbortController();

  task.subscribe('completed', () => ctrl.abort());

  // Stream
  let current = '';

  try {
    for await (const chunk of streamEvents(task, `stream.${stream}`, { signal: ctrl.signal })) {
      const data = current + chunk.data.toString('utf-8');
      const lines = data.split(/\r?\n/g);

      current = lines.pop() ?? '';

      for (const line of lines) {
        yield line;
      }
    }
  } catch (err) {
    if (task.exitCode !== 0) {
      throw err;
    }

    if (current) {
      yield current;
    }
  }
}

// Command utils
export interface Middleware<T = unknown, U = unknown> {
  builder?: (yargs: Argv<T>) => Argv<U>;
  handler(args: Arguments<U>): Awaitable<void>;
}

export function applyMiddlewares<T>(yargs: Argv<T>, middlewares: Middleware[]): Argv<T> {
  let tmp: Argv<unknown> = yargs;

  for (const middleware of middlewares) {
    if (middleware.builder) {
      tmp = middleware.builder(tmp);
    }

    tmp.middleware(middleware.handler);
  }

  return tmp as Argv<T>;
}

export function defineCommand<T, U>(command: CommandModule<T, U>): CommandModule<T, U> {
  return command;
}

export function defineMiddleware<T, U>(middleware: Middleware<T, U>): Middleware<T, U> {
  return middleware;
}

// String utils
export function capitalize(txt: string): string {
  return txt.charAt(0).toUpperCase() + txt.substring(1).toLowerCase();
}
