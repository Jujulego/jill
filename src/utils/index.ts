export * from './import';
export * from './streams';
export * from './string';
export * from './yargs';
