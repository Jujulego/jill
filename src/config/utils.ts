import { interfaces as int } from 'inversify';
import { ValidateFunction } from 'ajv';
import { cosmiconfig, defaultLoaders } from 'cosmiconfig';

import { AJV } from '@/src/ajv.config';
import schema from '@/src/assets/schema.json';
import { container } from '@/src/inversify.config';
import { dynamicImport } from '@/src/utils/import';

import { type IConfig, type IConfigExplorer } from './types';

// Symbols
export const CONFIG_EXPLORER: int.ServiceIdentifier<IConfigExplorer> = Symbol('jujulego:jill:config-explorer');
export const CONFIG_VALIDATOR: int.ServiceIdentifier<ValidateFunction<IConfig>> = Symbol('jujulego:jill:config-validator');

// Setup
container
  .bind(CONFIG_VALIDATOR)
  .toDynamicValue((context) => {
    const ajv = context.container.get(AJV);
    return ajv.compile<IConfig>(schema);
  })
  .inSingletonScope();

container
  .bind(CONFIG_EXPLORER)
  .toDynamicValue(() => {
    return cosmiconfig('jill', {
      loaders: {
        '.cjs': (filepath) => dynamicImport(filepath).then((mod) => mod.default),
        '.js': (filepath) => dynamicImport(filepath).then((mod) => mod.default),
        '.json': defaultLoaders['.json'],
        '.yaml': defaultLoaders['.yaml'],
        '.yml': defaultLoaders['.yml'],
        noExt: defaultLoaders.noExt,
      }
    });
  })
  .inSingletonScope();
