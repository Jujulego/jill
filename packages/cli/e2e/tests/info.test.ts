import chalk from 'chalk';
import * as path from 'path';

import { jill, MOCK } from '../utils';

// Setup
chalk.level = 1;

// Tests suites
describe('jill info', () => {
  it('should print main workspace data', async () => {
    await expect(jill(['info'], { cwd: MOCK })).
      resolves.toMatchSnapshot();
  });

  it('should print given workspace data', async () => {
    await expect(jill(['info', '-w', 'mock-test-a'], { cwd: MOCK })).
      resolves.toMatchSnapshot();
  });

  it('should print current workspace data', async () => {
    await expect(jill(['info'], { cwd: path.join(MOCK, 'workspaces/test-a') })).
      resolves.toMatchSnapshot();
  });
});