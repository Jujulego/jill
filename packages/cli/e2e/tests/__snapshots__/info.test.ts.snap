// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jill info should print current workspace data 1`] = `
Object {
  "code": 0,
  "stderr": Array [
    "- Loading project",
  ],
  "stdout": Array [
    "Workspace [1mmock-test-a[22m:",
    "[1mVersion:[22m   undefined",
    "[1mDirectory:[22m .",
    "",
    "[1mDependencies:[22m",
    "- mock-test-b",
    "",
    "[1mDev-Dependencies:[22m",
    "- mock-test-c",
  ],
}
`;

exports[`jill info should print given workspace data 1`] = `
Object {
  "code": 0,
  "stderr": Array [
    "- Loading project",
  ],
  "stdout": Array [
    "Workspace [1mmock-test-a[22m:",
    "[1mVersion:[22m   undefined",
    "[1mDirectory:[22m workspaces\\\\test-a",
    "",
    "[1mDependencies:[22m",
    "- mock-test-b",
    "",
    "[1mDev-Dependencies:[22m",
    "- mock-test-c",
  ],
}
`;

exports[`jill info should print main workspace data 1`] = `
Object {
  "code": 0,
  "stderr": Array [
    "- Loading project",
  ],
  "stdout": Array [
    "Workspace [1mmock-root[22m:",
    "[1mVersion:[22m   undefined",
    "[1mDirectory:[22m .",
  ],
}
`;
